### PR TITLE
crimson/os/seastore: implement label-based profiling with metrics

### PR DESCRIPTION
**Pull request by cyx1231st**
TODOs:
* Cache hit;
* Disk I/O;
* Transaction usage;

After 15 seconds rados bench write:
```
    "cache_trans_committed(shard=0, src=ALL)": 1322,
    "cache_trans_committed(shard=0, src=CLEANER)": 1,
    "cache_trans_committed(shard=0, src=SEASTORE)": 1321,
    "cache_trans_committed(shard=0, src=TEST)": 0,
    "cache_trans_created(shard=0, src=ALL)": 5175,
    "cache_trans_created(shard=0, src=CLEANER)": 1,
    "cache_trans_created(shard=0, src=SEASTORE)": 3469,
    "cache_trans_created(shard=0, src=SEASTORE_READ)": 1704,
    "cache_trans_created(shard=0, src=TEST)": 0,
    "cache_trans_created(shard=0, src=WEAK)": 1,
    "cache_trans_invalidated(ext=COLL_BLOCK, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=COLL_BLOCK, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=COLL_BLOCK, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=COLL_BLOCK, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=LADDR_INTERNAL, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=LADDR_INTERNAL, shard=0, src=SEASTORE)": 62,
    "cache_trans_invalidated(ext=LADDR_INTERNAL, shard=0, src=SEASTORE_READ)": 19,
    "cache_trans_invalidated(ext=LADDR_INTERNAL, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=LADDR_LEAF, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=LADDR_LEAF, shard=0, src=SEASTORE)": 1121,
    "cache_trans_invalidated(ext=LADDR_LEAF, shard=0, src=SEASTORE_READ)": 406,
    "cache_trans_invalidated(ext=LADDR_LEAF, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=OBJECT_DATA_BLOCK, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=OBJECT_DATA_BLOCK, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=OBJECT_DATA_BLOCK, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=OBJECT_DATA_BLOCK, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=OMAP_INNER, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=OMAP_INNER, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=OMAP_INNER, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=OMAP_INNER, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=ONODE_BLOCK_STAGED, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=ONODE_BLOCK_STAGED, shard=0, src=SEASTORE)": 368,
    "cache_trans_invalidated(ext=ONODE_BLOCK_STAGED, shard=0, src=SEASTORE_READ)": 40,
    "cache_trans_invalidated(ext=ONODE_BLOCK_STAGED, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=RBM_ALLOC_INFO, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=RBM_ALLOC_INFO, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=RBM_ALLOC_INFO, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=RBM_ALLOC_INFO, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=RETIRED_PLACEHOLDER, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=RETIRED_PLACEHOLDER, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=RETIRED_PLACEHOLDER, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=RETIRED_PLACEHOLDER, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=ROOT, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=ROOT, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=ROOT, shard=0, src=SEASTORE_READ)": 1,
    "cache_trans_invalidated(ext=ROOT, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK_PHYSICAL, shard=0, src=CLEANER)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK_PHYSICAL, shard=0, src=SEASTORE)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK_PHYSICAL, shard=0, src=SEASTORE_READ)": 0,
    "cache_trans_invalidated(ext=TEST_BLOCK_PHYSICAL, shard=0, src=TEST)": 0,
    "cache_trans_invalidated(shard=0, src=ALL)": 2017,
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/admin/osd_admin.cc [hunk 1]
```diff
@@ -248,40 +248,44 @@ class DumpMetricsHook : public AdminSocketHook {
...
-        f->dump_string("le", "+Inf");
+    try {
+      switch (auto v = metric(); v.type()) {
+      case data_type::GAUGE:
+        f->dump_float(name, v.d());
+        break;
+      case data_type::COUNTER:
+        [[fallthrough]];
+      case data_type::DERIVE:
+        f->dump_unsigned(name, v.ui());
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:259]: @cyx1231st i think we should fix it instead, see #42203

Comment (cyx1231st) [reply src/crimson/admin/osd_admin.cc:259]: Oh, thanks, should we keep the try-catch as a fall-back to dump metircs?

Comment (tchaikov) [reply src/crimson/admin/osd_admin.cc:259]: no, i'd just crash the application. as i think the mapping from datatype and the underlying type in the `variant<>` is a contract between the application (crimson) and the library (seastar). if the contract is broken, that's a bug.

File: src/crimson/os/seastore/cache.h [hunk 2]
```diff
@@ -94,30 +94,23 @@ class Cache {
...
       last_commit
     );
     retired_extent_gate.add_token(ret->retired_gate_token);
-    DEBUGT("created", *ret);
+    DEBUGT("created source={}", *ret, src);
     return ret;
   }
 
-  /// Creates empty weak transaction
+  /// Creates empty weak transaction (test usage only)
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/cache.h:115]: this is not only for test, mount also need weak transaction.

File: src/crimson/os/seastore/transaction.h [hunk 3]
```diff
@@ -133,8 +133,24 @@ class Transaction {
     return retired_set;
   }
 
+  enum class src_t {
```
Comment (tchaikov) [inline src/crimson/os/seastore/transaction.h:144]: nit,
```c++
enum class src_t : std::uint8_t
```

File: src/crimson/admin/osd_admin.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/admin/osd_admin.cc]: commit dropped.

File: src/crimson/os/seastore/transaction.h [hunk 4]
```diff
@@ -191,11 +207,7 @@ class Transaction {
   friend class Cache;
   friend Ref make_test_transaction();
 
-  /**
-   * If set, *this may not be used to perform writes and will not provide
-   * consistentency allowing operations using to avoid maintaining a read_set.
-   */
-  const bool weak;
+  const src_t src;
```
Comment (tchaikov) [inline src/crimson/os/seastore/transaction.h:217]: nit, might want to rearrange the layout of this struct for smaller memory footprint.

File: src/crimson/os/seastore/transaction.h [hunk 5]
```diff
@@ -141,8 +141,24 @@ class Transaction {
     return retired_set;
   }
 
+  enum class src_t {
+    SEASTORE,
+    SEASTORE_READ,
```
Comment (tchaikov) [inline src/crimson/os/seastore/transaction.h:146]: at first glance, i'd say `SEASTORE` is a superset of `SEASTORE_READ`. if that's the case, why not s/SEASTORE/SEASTORE_WRITE/ or s/SEASTORE/SEASTORE_MUTATE/?

**Comment by liu-chunmei** [comment]
@cyx1231st what command to dump these metrics?

File: src/crimson/os/seastore/cache.h [hunk 6]
```diff
@@ -550,6 +558,12 @@ class Cache {
    */
   CachedExtent::list dirty;
 
+  struct {
+    std::map<Transaction::src_t, uint64_t> trans_created;
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.h:562]: i'd prefer `vector<>` if possible.

File: src/crimson/os/seastore/cache.cc [hunk 7]
```diff
@@ -80,6 +83,67 @@ void Cache::dump_contents()
...
+      }
+    );
+  };
+  for (auto& src : {src_t::SEASTORE,
+                    src_t::SEASTORE_READ,
+                    src_t::CLEANER,
+                    src_t::TEST,
+                    src_t::WEAK}) {
+    f_register_trans_created(src);
+  }
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.cc:141]: nit,
```c++
for (auto& [src, label] : labels_by_src) {
  f_register_trans_created(src, label);
}
```

File: src/crimson/os/seastore/cache.cc [hunk 8]
```diff
@@ -80,6 +83,67 @@ void Cache::dump_contents()
...
+    {src_t::SEASTORE_READ, src_label("SEASTORE_READ")},
+    {src_t::CLEANER,       src_label("CLEANER")},
+    {src_t::TEST,          src_label("TEST")},
+    {src_t::WEAK,          src_label("WEAK")}
+  };
+
+  /*
+   * trans_created
+   */
+  auto f_register_trans_created = [this, &labels_by_src](src_t src) {
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.cc:118]: nit, s/f_register_trans_created/register_trans_created/

**Comment by tchaikov** [comment]
> @cyx1231st what command to dump these metrics?

@liu-chunmei i am documenting some approaches to collect the metrics.

File: src/crimson/os/seastore/cache.h (hunk 6 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:562]: Do you think `unordered_map<>` is faster for the hot-path to increase the counter indexed by the related labels?

File: src/crimson/os/seastore/cache.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:115]: After this change only cache unit test will call this interface.

**Comment by cyx1231st** [comment]
> @cyx1231st what command to dump these metrics?

I used `ceph daemon <socket-addr> dump_metrics`

**Comment by tchaikov** [comment]
> > @cyx1231st what command to dump these metrics?
> 
> @liu-chunmei i am documenting some approaches to collect the metrics.

#42205

File: src/crimson/os/seastore/transaction.h (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h]: fixed.

File: src/crimson/os/seastore/transaction.h (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h]: moved to the last

File: src/crimson/os/seastore/transaction.h (hunk 5 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h]: renamed to `SEASTORE_MUTATE`

File: src/crimson/os/seastore/cache.h (hunk 6 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: changed to `unordered_map<>`

File: src/crimson/os/seastore/cache.cc (hunk 7 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc]: fixed

File: src/crimson/os/seastore/cache.cc (hunk 8 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc]: done.

File: src/crimson/os/seastore/cache.h (hunk 6 above)
Comment (tchaikov) [reply src/crimson/os/seastore/cache.h]: i think the distribution of `src_t` won't be very sparse. actually, they are but consecutive integers, so i think vector is much more spacial-efficient than a red-black tree.

File: src/crimson/admin/osd_admin.cc [hunk 9]
```diff
@@ -231,20 +231,35 @@ class DumpMetricsHook : public AdminSocketHook {
 
   static void dump_metric_value(Formatter* f,
                                 string_view full_name,
-                                const registered_metric& metric)
+                                const registered_metric& metric,
+                                const seastar::metrics::impl::labels_type& labels)
   {
+    std::ostringstream oss_name;
+    oss_name << full_name << "(";
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:238]: if we really want to leverage labels as part of the naming scheme, i'd suggest use a more structured approach.

1. to support syntax like `ceph dump-metrics metric_name label_name_a=label_value_a label_name_b=label_value_b...`, so we can query and print a certain metric with specified label kv pairs.
2. dump each metric like an object, like `{"value": 123, "label_name_a": "label_value_a", ...}`. so one is able to pipe the output to `jq` for further processing

File: src/crimson/os/seastore/cache.h (hunk 6 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: Yeah, to index consecutive integers, vector is better. I'll change trans_created/committed to vectors.

File: src/crimson/admin/osd_admin.cc (hunk 9 above)
Comment (cyx1231st) [reply src/crimson/admin/osd_admin.cc:238]: I'll modify the dump format to fit the structure as suggested, for example.
`cache_trans_committed{"shard": 0, "src": "CLEANER", "value": 1}", ...`

File: src/crimson/os/seastore/cache.h (hunk 6 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: see https://github.com/ceph/ceph/pull/42202/commits/295c5c34ad3573667b6c4e136376708ea07bffff

File: src/crimson/admin/osd_admin.cc (hunk 9 above)
Comment (cyx1231st) [reply src/crimson/admin/osd_admin.cc]: see https://github.com/ceph/ceph/pull/42202/commits/6b38bf1f1be71238d45185433c066b4476c1f500

File: src/crimson/os/seastore/cache.h [hunk 10]
```diff
@@ -562,8 +562,8 @@ class Cache {
 
   using src_ext_t = std::pair<Transaction::src_t, extent_types_t>;
   struct {
-    std::unordered_map<Transaction::src_t, uint64_t> trans_created;
-    std::unordered_map<Transaction::src_t, uint64_t> trans_committed;
+    std::vector<uint64_t> trans_created_by_src;
+    std::vector<uint64_t> trans_committed_by_src;
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.h:566]: could have simple helper methods here for indexing these vectors using `src_t`.

**Review by tchaikov — APPROVED** [review]
probably we can have "crimson/common: dump metric with labels" into its own PR. so it can be merged sooner. regarding other changes, i think they do make sense to me. but i'd like see what the team think about it.

File: src/crimson/os/seastore/cache.h (hunk 2 above)
Comment (athanatos) [reply src/crimson/os/seastore/cache.h:117]: Remove the (test usage only) portion.  I interpret it to mean that weak transactions aren't used outside of tests which is not true.  Probably should use this helper in TransactionManager::mount() anyway.

File: src/crimson/os/seastore/transaction.h [hunk 11]
```diff
@@ -141,8 +141,25 @@ class Transaction {
...
+    SEASTORE_MUTATE = 0,
+    SEASTORE_READ,
+    CLEANER,
+    // including tests and tools that bypass seastore
+    TEST,
+    /**
+     * If set, *this may not be used to perform writes and will not provide
+     * consistentency allowing operations using to avoid maintaining a read_set.
+     */
+    WEAK,
```
Comment (athanatos) [inline src/crimson/os/seastore/transaction.h:154]: WEAK is orthogonal to source.  Remove from src_t and add the weak boolean and helpers back in.  Probably worth adding an INIT src_t value for mkfs/mount though.

File: src/crimson/os/seastore/transaction.h [hunk 12]
```diff
@@ -141,8 +141,25 @@ class Transaction {
     return retired_set;
   }
 
+  enum class src_t : uint8_t {
+    SEASTORE_MUTATE = 0,
+    SEASTORE_READ,
+    CLEANER,
+    // including tests and tools that bypass seastore
+    TEST,
```
Comment (athanatos) [inline src/crimson/os/seastore/transaction.h:149]: I also don't really like TEST.  The same scenario won't have both TEST and MUTATE (for instance) operations, so distinguishing between them isn't useful.  Distinguishing between MUTATE, READ, and CLEANER within a test, however, is potentially useful.

File: src/crimson/os/seastore/transaction.h (hunk 11 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h:154]: The current `SEASTORE_READ` is categorized as seastore-level read operations that expect conflicts. This is why we can calculate the accurate conflict rate from equation `trans_invalidated[SEASTORE_READ] / trans_created[SEASTORE_READ]`.

Strictly speaking we need another orthogonal label maybe called "op_type_t" including "READ", "WRITE", and "WEAK", and "src_t" would include "TEST", "SEASTORE", "CLEANER" and "INIT". But that would be too many and some combinations don't exist or not meaningful for performance profiling (mostly unit tests), that's why I combined them into the existing 5 categories.

File: src/crimson/os/seastore/transaction.h (hunk 12 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h:149]: see https://github.com/ceph/ceph/pull/42202#discussion_r666634114

File: src/crimson/os/seastore/cache.h (hunk 10 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:564]: introduced `get_counter()` helper.

**Comment by cyx1231st** [comment]
The currrent `src_t` is a combined label of two orthogonal categories and this is the mapping:
*  INIT - WEAK  ========> WEAK
*  INIT - READ (doesn't exist)
*  INIT - MUTATE =======> SEASTORE_MUTATE
*  SEASTORE - WEAK (doesn't exist)
*  SEASTORE - READ =====> SEASTORE_READ
*  SEASTORE - MUTATE ===> SEASTORE_MUTATE
*  CLEANER - WEAK (doesn't exist)
*  CLEANER - READ (doesn't exist)
*  CLEANER - MUTATE ====> CLEANER
*  TEST - WEAK ========> WEAK
*  TEST - READ =========> TEST
*  TEST - MUTATE =======> TEST

Please help suggest better ones if the existing `src_t` is not appropriate enough.

File: src/crimson/os/seastore/cache.h [hunk 13]
```diff
@@ -562,11 +560,17 @@ class Cache {
 
   using src_ext_t = std::pair<Transaction::src_t, extent_types_t>;
   struct {
-    std::unordered_map<Transaction::src_t, uint64_t> trans_created;
-    std::unordered_map<Transaction::src_t, uint64_t> trans_committed;
+    std::vector<uint64_t> trans_created_by_src;
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.h:563]: after a second thought, `std::array<uint64_t, static_cast<size_t>(Transaction::src_t::MAX)>` would be more appropriate. as the sizes of these containers are known at compile time.

Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:563]: I think size consideration is not a big matter here as Cache and its stats are singleton. And we already resize the vector to a proper size during initialization ...

Comment (tchaikov) [reply src/crimson/os/seastore/cache.h:563]: agreed. it's not a big deal from the performance perspective. it's just "more appropriate and simpler". and i don't feel strong either way.

FWIW, resizing vector does not change its capacity in some cases, see http://en.cppreference.com/w/cpp/container/vector/resize#Notes

Comment (ronen-fr) [reply src/crimson/os/seastore/cache.h:563]: I'd vote with Kefu on that: we should use std::array where possible.

Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: see https://github.com/ceph/ceph/pull/42202/commits/8a5bbd2a79db43cf41d012ad7260894f4d933c75

File: src/crimson/os/seastore/cache.h [hunk 14]
```diff
@@ -94,34 +94,29 @@ class Cache {
...
-    DEBUGT("created", *ret);
+    DEBUGT("created source={}", *ret, src);
     return ret;
   }
 
   /// Resets transaction preserving
   void reset_transaction_preserve_handle(Transaction &t) {
+    if (t.did_reset()) {
+      assert(stats.trans_created.count(t.get_src()));
+      ++(stats.trans_created[t.get_src()]);
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: Fixed by https://github.com/ceph/ceph/pull/42202/commits/42fcd59fabb01a5df54551f3729fdb33a831cb43, now the above equation is even.

File: src/crimson/os/seastore/cache.h [hunk 15]
```diff
@@ -281,9 +284,14 @@ class Cache {
       return trans_intr::make_interruptible(
 	get_extent<T>(offset, length)
       ).si_then(
-	[&t](auto ref) mutable {
+	[&t, this](auto ref) mutable {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.h:287]: remove `mutable`?

File: src/crimson/os/seastore/transaction.h (hunk 11 above)
Comment (athanatos) [reply src/crimson/os/seastore/transaction.h:154]: WEAK is irrelevant to profiling.  INIT would be the op type as it refers to mkfs and mount.  I want op_type_t to be purely informational and do not want it to impact behavior.  Please restore the weak boolean and related behavior.

File: src/crimson/os/seastore/transaction.h (hunk 12 above)
Comment (athanatos) [reply src/crimson/os/seastore/transaction.h:149]: I disagree.  I don't see any reason why opertations originating from tests can't be tagged as READ/WRITE/CLEANER.

**Comment by athanatos** [comment]
> The currrent `src_t` is a combined label of two orthogonal categories and this is the mapping:
> 
>     * INIT - WEAK  ========> WEAK
> 
>     * INIT - READ (doesn't exist)
> 
>     * INIT - MUTATE =======> SEASTORE_MUTATE
> 
>     * SEASTORE - WEAK (doesn't exist)
> 
>     * SEASTORE - READ =====> SEASTORE_READ
> 
>     * SEASTORE - MUTATE ===> SEASTORE_MUTATE
> 
>     * CLEANER - WEAK (doesn't exist)
> 
>     * CLEANER - READ (doesn't exist)
> 
>     * CLEANER - MUTATE ====> CLEANER
> 
>     * TEST - WEAK ========> WEAK
> 
>     * TEST - READ =========> TEST
> 
>     * TEST - MUTATE =======> TEST
> 
> 
> Please help suggest better ones if the existing `src_t` is not appropriate enough.

Let's do:

     * INIT - READ =======> INIT
     * INIT - MUTATE =======> INIT
     * OPERATION - READ =====> SEASTORE_READ
     * OPERATION - MUTATE ===> SEASTORE_MUTATE
     * CLEANER - READ ====> CLEANER
     * CLEANER - MUTATE ====> CLEANER

OPERTATION - READ/MUTATE refer to normal IO operations whether they originate at the seastore.h boundary or within a test.

WEAK actually influences behavior and I really don't want it included as part of src_t -- readers would find that very confusing.

**Comment by cyx1231st** [comment]
```
* INIT - READ =======> INIT
 * INIT - MUTATE =======> INIT
 * OPERATION - READ =====> SEASTORE_READ
 * OPERATION - MUTATE ===> SEASTORE_MUTATE
 * CLEANER - READ ====> CLEANER
 * CLEANER - MUTATE ====> CLEANER
```

Make sense to me, working on it.

File: src/crimson/os/seastore/cache.h (hunk 15 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: removed.

File: src/crimson/os/seastore/transaction.h (hunk 11 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h]: reverted the weak boolean in https://github.com/ceph/ceph/pull/42202/commits/312340317e1f8bda31a9b27ec4ea7c3396dcb088

**Comment by cyx1231st** [comment]
change-set:
* crimson/os/seastore: revert to bool Transaction::weak and drop src_t::WEAK
* crimson/os/seastore: improve transaction source classification

**Comment by athanatos** [comment]
Merge the review fixes into the original commits -- that'll make it easy to confirm that the weak changes etc got reverted fully.

File: src/crimson/os/seastore/seastore.cc [hunk 16]
```diff
@@ -84,7 +84,7 @@ seastar::future<> SeaStore::mkfs(uuid_d new_osd_fsid)
   }).safe_then([this] {
     return seastar::do_with(
       transaction_manager->create_transaction(
-        Transaction::src_t::SEASTORE_MUTATE),
+        Transaction::src_t::MUTATE),
```
Comment (cyx1231st) [inline src/crimson/os/seastore/seastore.cc:87]: Is it MUTATE or INIT?

**Comment by cyx1231st** [comment]
> Merge the review fixes into the original commits -- that'll make it easy to confirm that the weak changes etc got reverted fully.

Merged all fixes.

**Review by athanatos — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Comment by xxhdx1985126** [comment]
It seems that we should also check the validity of extents and count invalidates [here](https://github.com/ceph/ceph/blob/40148fc24819642a1f087a5d3a03f679fe86ed33/src/crimson/os/seastore/cache.h#L294), am I right?

**Comment by xxhdx1985126** [comment]
and [here](https://github.com/ceph/ceph/blob/40148fc24819642a1f087a5d3a03f679fe86ed33/src/crimson/os/seastore/cache.h#L346)

**Comment by cyx1231st** [comment]
> It seems that we should also check the validity of extents and count invalidates here, am I right?
>
> and here

IIUC it is not necessary because this is what interruptive-future is already doing -- when an extent is invalidated and it is in other transaction's read-set, those transactions will be marked invalidated, and they will be interrupted by `si_then()` as soon as the continuations of those transactions are resumed.

**Pull request merged**